Regression after recent commits.

After pulling from the latest master, I can't use lambda's as member types:

```
out/main.c:10088:36: error: initializing 'void *' with an expression of incompatible type 'Lambda'
        Lambda _13 = { .callback = f, .env = NULL, .delete = NULL, .copy = NULL }; //Sym f Symbol
```

I've confirmed that there's no issue as of commit de8104eb--the problem arose someplace after that commit.

--- a/src/Qualify.hs
+++ b/src/Qualify.hs
@@ -96,8 +96,6 @@ setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst (matchExpr@(XObj Match
        in  XObj (Lst (matchExpr : newExpr : concat newCasesXObjs)) i t
   else XObj (Lst (matchExpr : expr : casesXObjs)) i t -- Leave it untouched for the compiler to find the error.
 
-
-
 setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [XObj With _ _, XObj (Sym path _) _ _, expression]) _ _) =
   let useThese = envUseModules env
       env' = if path `elem` useThese then env else env { envUseModules = path : useThese }

--- a/src/Qualify.hs
+++ b/src/Qualify.hs
@@ -73,10 +73,23 @@ setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst (matchExpr@(XObj Match
            innerEnv = Env Map.empty (Just env) (Just ("case-env-" ++ show (infoIdentifier ii))) [] InternalEnv lvl
            newCasesXObjs =
              map (\(l, r) ->
-                    let l' = setFullyQualifiedSymbols typeEnv globalEnv env l
-                        -- NOTE: Used to extend innerEnv with bindings from the lhs here. Doesn't seem necessary, and cleans up the code not to do it.
-                        r' = setFullyQualifiedSymbols typeEnv globalEnv innerEnv r
-                    in  [l', r']
+                    case l of
+                      XObj (Lst (x:xs)) _ _ ->
+                        let l' = setFullyQualifiedSymbols typeEnv globalEnv env l
+                            innerEnv' = foldl' (\e v ->
+                                                  case v of
+                                                    XObj (Sym (SymPath _ binderName) _) _ _ ->
+                                                      extendEnv e binderName v
+                                                    x ->
+                                                      error ("Can't match variable with " ++ show x))
+                                               innerEnv
+                                               xs
+                            r' = setFullyQualifiedSymbols typeEnv globalEnv innerEnv' r
+                        in  [l', r']
+                      XObj{} ->
+                        let l' = setFullyQualifiedSymbols typeEnv globalEnv env l
+                            r' = setFullyQualifiedSymbols typeEnv globalEnv env r
+                        in  [l', r']
                  ) (pairwise casesXObjs)
        in  XObj (Lst (matchExpr : newExpr : concat newCasesXObjs)) i t
   else XObj (Lst (matchExpr : expr : casesXObjs)) i t -- Leave it untouched for the compiler to find the error.

--- a/src/Qualify.hs
+++ b/src/Qualify.hs
@@ -76,14 +76,16 @@ setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst (matchExpr@(XObj Match
                     case l of
                       XObj (Lst (x:xs)) _ _ ->
                         let l' = setFullyQualifiedSymbols typeEnv globalEnv env l
-                            innerEnv' = foldl' (\e v ->
-                                                  case v of
-                                                    XObj (Sym (SymPath _ binderName) _) _ _ ->
-                                                      extendEnv e binderName v
-                                                    x ->
-                                                      error ("Can't match variable with " ++ show x))
-                                               innerEnv
-                                               xs
+                            innerEnv' = foldl' folder innerEnv xs
+                                      where folder e v = case v of
+                                                       XObj (Sym (SymPath _ binderName) _) _ _ ->
+                                                         extendEnv e binderName v
+                                                       -- Nested sumtypes
+                                                       -- fold recursively -- is there a more efficient way?
+                                                       XObj (Lst(y:ys)) _ _ ->
+                                                         foldl' folder innerEnv ys
+                                                       x ->
+                                                         error ("Can't match variable with " ++ show x)
                             r' = setFullyQualifiedSymbols typeEnv globalEnv innerEnv' r
                         in  [l', r']
                       XObj{} ->
@@ -93,6 +95,8 @@ setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst (matchExpr@(XObj Match
                  ) (pairwise casesXObjs)
        in  XObj (Lst (matchExpr : newExpr : concat newCasesXObjs)) i t
   else XObj (Lst (matchExpr : expr : casesXObjs)) i t -- Leave it untouched for the compiler to find the error.
+
+
 
 setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [XObj With _ _, XObj (Sym path _) _ _, expression]) _ _) =
   let useThese = envUseModules env

--- a/test/sumtypes.carp
+++ b/test/sumtypes.carp
@@ -1,0 +1,34 @@
+;; Generic tests on sumtypes.
+
+(load "Test.carp")
+(use Test)
+
+(deftype A (F [(Fn [] Int)]))
+(deftype (Nest a) (Nested [a]))
+
+(def nest (Nest.Nested (Nest.Nested (Nest.Nested 123))))
+
+(defn one-two-three [] 123)
+
+(defn m [a]
+  (match a
+    (A.F f) (f)))
+
+(deftest test 
+  (assert-equal test
+    123
+    (m (A.F one-two-three))
+    "match adds lhs bindings to inner environments")
+  (assert-equal test
+    "matched!"    
+    (match nest 
+          (Nest.Nested (Nest.Nested (Nest.Nested _)))
+          "matched!")
+    "Match matches nested sumtype constructors with underscores")
+  (assert-equal test
+    123    
+    (match nest 
+          (Nest.Nested (Nest.Nested (Nest.Nested x)))
+          x)
+    "Match matches nested sumtype constructors with variables")
+)

--- a/test/sumtypes.carp
+++ b/test/sumtypes.carp
@@ -4,7 +4,7 @@
 (use Test)
 
 (deftype A (F [(Fn [] Int)]))
-(deftype (Nest a) (Nested [a]))
+(deftype (Nest a) (Nested [a]) (End []))
 
 (def nest (Nest.Nested (Nest.Nested (Nest.Nested 123))))
 
@@ -21,7 +21,11 @@
     "match adds lhs bindings to inner environments")
   (assert-equal test
     "matched!"    
-    (match nest 
+    (match nest
+          (Nest.Nested (Nest.End))
+          "wrong!"
+          (Nest.Nested (Nest.Nested (Nest.End)))
+          "Also wrong!"
           (Nest.Nested (Nest.Nested (Nest.Nested _)))
           "matched!")
     "Match matches nested sumtype constructors with underscores")

